### DISCUSSION TITLE
Security

### DISCUSSION THREAD

**Discussion by GGAlanSmithee** [feldmanchain/whitepaper#6] [General]
# Security

There has come up a couple of security concerns in early discussions, so let's have a conversation regarding how to secure the feldmanchain. What follows is a (incomplete and ever growing) list of potential security concerns

1. How can I, as a project author, be sure that my code is not tampered with by a builder
1. How can I, as a project author, be sure that my code is not stolen by a builder
1. How can I, as a builder, be sure that the code I am building will not corrupt my computer, or othervise cause harm

Let's keep adding potential security concerns as they cmoe up, and hopefully potential solutions to them.

**Comment by GGAlanSmithee** [comment]
In regards to point 1 - **How can I, as a project author, be sure that my code is not tampered with by a builder** - I was thinking that we could build a mechanics into the feldmanchain that does either a full, or one or more partial, builds on _n_ nodes, where _n_ is some number that could possibly be affected by the project author.

The nodes to carry out the build would be randomly selected, to guarantee that there, with sufficient decentralization and a high enough _n_, statistically cannot be a coordinated effort across the selected nodes. After the builds are done, we could compare the checksums of the output, and reject if not matched.

This would introduce inefficiencies in tradeoff for security, but if we make it configurable, this could practically be disabled by setting n=1 in, for example, the situation when you are running the feldmanchain on prem in a private and secure network.

Thoughts @brokenprogrammer ?

EDIT: You are of course more than welcome to share your thoughts @JEPPSER (and anyone else that happens to find this or any other discussion)

**Reply by brokenprogrammer** [reply]
Having multiple runners performing the same job is one idea to validate the build however the idea that I had in mind was to do something similar to what other CI/CD systems do which is allowing a sort of scriptable "build job". 

Looking at Github workflows which is very similar if not the same to how azure pipelines work they allow you to use something called "Actions" to build your build script. For example this is a repository for the Action of checking out a repository: https://github.com/actions/checkout

One way could be to limit the supported "Actions" to ones that have been officially approved. I am not 100% sure about this idea yet but I am thinking more about it.

> How can I, as a project author, be sure that my code is not stolen by a builder

I think to battle this we would need to somehow be able to check if a runner that accepts a job has been compromised or not. As of now I do not know how to do this in a good way. Perhaps if we sign the application with our own keys when making a release and somehow make that play a role when getting a build assigned to you?

> How can I, as a builder, be sure that the code I am building will not corrupt my computer, or othervise cause harm

The end user's build environments would have to be a clean and isolated sandbox or virtual machine. I know github for example gives no guarantees when self-hosting their runners which makes sense.  Maybe adding additional limits to what the build scripts can do would help as well?

**Reply by GGAlanSmithee** [reply]
> Looking at Github workflows which is very similar if not the same to how azure pipelines work they allow you to use something called "Actions" to build your build script. For example this is a repository for the Action of checking out a repository: https://github.com/actions/checkout
>
> One way could be to limit the supported "Actions" to ones that have been officially approved. I am not 100% sure about this idea yet but I am thinking more about it.

I do like the idea of haiving the builder be GIT based, building of commits and running actions according to some configuration. This seems pretty standard and seem to work well in other setups.

In regards to limiting actions, that might be something worth looking at, but I am having a hard time seeing how we could possibly restrict someone from simply:

1. pull the code
1. make local changes (introducing malicious code)
1. building the project

Even if we somehow add restrictions for this, such that you cannot make a build if you have locally checked out files, or if you are not on the correct commit etc, what is stopping the attacker from simply copying the project, doing whatever alterations, building it and replacing the output, or simply change the output after the fact?

I think the attack surface is really large here, and I can foresee it being hard to give guarantees in this regard, which is why I thought of the idea of not relying on the software, but rather the system as a whole for security in regards to a statistical guarantee, since I think doing it on a technical level will prove to be very hard.

> > How can I, as a project author, be sure that my code is not stolen by a builder
>
> I think to battle this we would need to somehow be able to check if a runner that accepts a job has been compromised or not. As of now I do not know how to do this in a good way. Perhaps if we sign the application with our own keys when making a release and somehow make that play a role when getting a build assigned to you?

Something like a cryptographic signature is actually a really good idea. I know too little about this to understand what level of protection that would give, and how that would be implemented. I am not keen on the idea of requiring that network participants verify their identity in a decentralized system though, but that should be able to be avoided.

Also, I do think that running it in a VM, such as you've talked about before, is a really good idea. We need to research this, but if we can deliver the runner / build node as a fully contained VM, restricting all interactions with the host, that should give sufficient guarantees in this respect. But again, I do not know if that is even possible, and just how secure that would be.

> > How can I, as a builder, be sure that the code I am building will not corrupt my computer, or othervise cause harm
> 
> The end user's build environments would have to be a clean and isolated sandbox or virtual machine. I know github for example gives no guarantees when self-hosting their runners which makes sense. Maybe adding additional limits to what the build scripts can do would help as well?

Exactly, I really think we should look hard into this. It _should_ be possible to run a self-contained, fully isolated VM/sandbox. If we do implement some form of signature verification, that could go a long way.

Really good input @brokenprogrammer thanks!

**Reply by GGAlanSmithee** [reply]
To rebute my own reasoning in regards to point 1, if we do succeed in implementing a self-contained and ditigally verified VM/sandbox node, that would probably take care of that too. It makes me wonder though why such a solution isn't readily available already. (or is there?)

Another thing to keep in mind is that a key component of a blockchain is trustlessness, which in essense means that the trust between the network participants is built on economic incentivis rather than a governing body. In a traditional blockchain, you are free to manipulate it whichever way you want, but if you cannot convince 51% (this varies across different chains) of the network participants that your version of the chain is the correct one, then it will have no real impact. This is another reason why we might want to explore the idea of decentralized trust.

**Reply by GGAlanSmithee** [reply]
to rebute my rebuttal 🗡️ it will probably be hard to restrict anyone from, for example, monitoring outgoing traffic on the network and do a man-in-the-middle attack.

Two concepts I would like to explore is multi-sig or aggregate cryptographics signatures as well as sealed envelope auctions as a way to both secure and potentially balance the tokenomics side of things.
